File: backend/src/schema.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

table! {
    regions (id) {
        id -> Text,
        name -> Text,
    }
}

table! {
    contests (year) {
        year -> Integer,
        location -> Nullable<Text>,
        gmaps -> Nullable<Text>,
        latitude -> Nullable<Float>,
        longitude -> Nullable<Float>,
        region -> Nullable<Text>,
    }
}

table! {
    users (id) {
        id -> Text,
        name -> Text,
        surname -> Text,
        birth -> Nullable<Text>,
        gender -> Nullable<Text>,
    }
}

table! {
    participations (user_id, contest_year) {
        user_id -> Text,
        contest_year -> Integer,
        position -> Nullable<Integer>,
        school -> Nullable<Text>,
        venue -> Nullable<Text>,
        region -> Nullable<Text>,
        medal -> Nullable<Text>,
        IOI -> Nullable<Bool>,
        score -> Nullable<Float>,
    }
}

joinable!(participations -> users(user_id));
joinable!(participations -> regions(region));
allow_tables_to_appear_in_same_query!(participations, users);
allow_tables_to_appear_in_same_query!(participations, regions);

table! {
    tasks (name, contest_year) {
        name -> Text,
        contest_year -> Integer,
        index -> Integer,
        max_score -> Nullable<Float>,
        title -> Text,
        link -> Nullable<Text>,
    }
}

table! {
    task_scores (task_name, contest_year, user_id) {
        task_name -> Text,
        contest_year -> Integer,
        user_id -> Text,
        score -> Nullable<Float>,
    }
}

joinable!(task_scores -> users(user_id));
allow_tables_to_appear_in_same_query!(task_scores, users);

File: frontend/lib/db/schema.ts
import { integer, primaryKey, real, sqliteTable, text } from "drizzle-orm/sqlite-core";

export const regions = sqliteTable("regions", {
  id: text().primaryKey(),
  name: text().notNull(),
});

export const users = sqliteTable("users", {
  id: text().primaryKey(),
  firstName: text("name"),
  lastName: text("surname").notNull(),
  username: text(),
});

export const contests = sqliteTable("contests", {
  year: integer().primaryKey(),
  location: text(),
  regionId: text("region"),
  gmaps: text(),
  latitude: real(),
  longitude: real(),
});

export const tasks = sqliteTable("tasks", {
  name: text().primaryKey(),
  contestYear: integer("contest_year")
    .notNull()
    .references(() => contests.year),
  idx: integer(),
  maxScorePossible: real("max_score_possible"),
  title: text(),
  link: text(),
});

export type Medal = "gold" | "silver" | "bronze";

export const participations = sqliteTable(
  "participations",
  {
    userId: text("user_id")
      .notNull()
      .references(() => users.id),
    contestYear: integer("contest_year")
      .notNull()
      .references(() => contests.year),
    rank: integer(),
    school: text(),
    regionId: text("region_id").references(() => regions.id),
    medal: text().$type<Medal>(),
    internationals: text(),
    score: real(),
  },
  (table) => [
    primaryKey({
      columns: [table.userId, table.contestYear],
      name: "participations_user_id_contest_year_pk",
    }),
  ],
);

export const taskScores = sqliteTable(
  "task_scores",
  {
    taskName: text("task_name")
      .notNull()
      .references(() => tasks.name),
    userId: text("user_id")
      .notNull()
      .references(() => users.id),
    contestYear: integer("contest_year")
      .notNull()
      .references(() => contests.year),
    score: real(),
  },
  (table) => [
    primaryKey({
      columns: [table.taskName, table.userId, table.contestYear],
      name: "task_scores_task_name_user_id_contest_year_pk",
    }),
  ],
);

export const internationals = sqliteTable("internationals", {
  code: text().primaryKey(),
  name: text().notNull(),
  link: text(),
  color: text(),
});
